jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -36,96 +36,108 @@
   /** Represents the different types of layers for a Java application. */
   @VisibleForTesting
   enum LayerType {
-    DEPENDENCIES("dependencies", JavaEntrypointConstructor.DEFAULT_DEPENDENCIES_PATH_ON_IMAGE),
-    SNAPSHOT_DEPENDENCIES(
-        "snapshot dependencies", JavaEntrypointConstructor.DEFAULT_DEPENDENCIES_PATH_ON_IMAGE),
-    RESOURCES("resources", JavaEntrypointConstructor.DEFAULT_RESOURCES_PATH_ON_IMAGE),
-    CLASSES("classes", JavaEntrypointConstructor.DEFAULT_CLASSES_PATH_ON_IMAGE),
+    DEPENDENCIES("dependencies"),
+    SNAPSHOT_DEPENDENCIES("snapshot dependencies"),
+    RESOURCES("resources"),
+    CLASSES("classes"),
     // TODO: remove this once we put files in WAR into the relevant layers (i.e., dependencies,
     // snapshot dependencies, resources, and classes layers). Should copy files in the right
-    EXPLODED_WAR("exploded war", JavaEntrypointConstructor.DEFAULT_JETTY_BASE_ON_IMAGE),
-    EXTRA_FILES("extra files", "/");
+    EXPLODED_WAR("exploded war"),
+    EXTRA_FILES("extra files");
 
     private final String name;
-    private final Path extractionPath;
 
-    /** Initializes with a name for the layer and the layer files' default extraction path root. */
-    LayerType(String name, String extractionPath) {
+    /**
+     * Initializes with a name for the layer and the layer files' default extraction path root.
+     *
+     * @param name name to set for the layer; does not affect the contents of the layer
+     */
+    LayerType(String name) {
       this.name = name;
-      this.extractionPath = Paths.get(extractionPath);
     }
 
     @VisibleForTesting
     String getName() {
       return name;
     }
-
-    @VisibleForTesting
-    Path getExtractionPath() {
-      return extractionPath;
-    }
   }
 
   /** Builds with each layer's files. */
   public static class Builder {
 
     private final Map<LayerType, List<Path>> layerFilesMap = new EnumMap<>(LayerType.class);
+    private final Map<LayerType, String> extractionPathMap = new EnumMap<>(LayerType.class);
 
     private Builder() {
       for (LayerType layerType : LayerType.values()) {
         layerFilesMap.put(layerType, new ArrayList<>());
+        extractionPathMap.put(layerType, "/");
       }
     }
 
-    public Builder setDependencyFiles(List<Path> dependencyFiles) {
+    public Builder setDependencyFiles(List<Path> dependencyFiles, String extractionPath) {
```

Comment:
Should perhaps `JavaLayerConfigurations` take an `appRoot` and build the `extractionPaths` rather than taking a separate `extractionPath` for each and delegating the `extractionPath` building to the caller?

Reply:
I think what you said is basically the original design, where the builder had `setAppRoot()` (https://github.com/GoogleContainerTools/jib/pull/984#discussion_r218169036)? The new design allows more flexibility that can cover the WAR case too. For example, the resources layer goes into "<appRoot>/resources" for non-WAR, but it should go to "<appRoot>" in the WAR case.

Reply:
Ah right.